### PR TITLE
add template for C# azure functions

### PR DESCRIPTION
**Pull request by watashiSHUN**
working on a generalization, in case that didn't go well (break stuff), this is fall back code ready to check in

### CODE REVIEW

File: Kudu.Core/Deployment/Generator/AspNetCoreBuilder.cs [hunk 1]
```diff
@@ -6,25 +6,23 @@ namespace Kudu.Core.Deployment.Generator
...
                 _version = "csproj";
             }
-            else if (_projectPath.EndsWith(".xproj", StringComparison.OrdinalIgnoreCase))
+            else if (_projectFilePath.EndsWith(".xproj", StringComparison.OrdinalIgnoreCase))
             {
                 // if it's xproj, throw invalidOperationException
-                throw new InvalidOperationException(String.Format(CultureInfo.CurrentCulture,
-                                                             Resources.Error_ProjectNotDeployable,
-                                                             projectPath));
+                throw new InvalidOperationException("Building Asp.Net Core .xproj is no longer supported in Azure, please move to .csproj");
```
Comment (davidebbo) [inline Kudu.Core/Deployment/Generator/AspNetCoreBuilder.cs]: Also add this link I just created for it in the error message: https://go.microsoft.com/fwlink/?linkid=850964

File: Kudu.Core/Deployment/Generator/FunctionAppBuilder.cs [hunk 2]
```diff
@@ -2,18 +2,38 @@
...
+            _solutionPath = solutionPath;
+        }
+
+        protected override string ScriptGeneratorCommandArguments
         {
+            get
+            {
+                if (string.IsNullOrEmpty(_solutionPath))
+                {
+                    return $"--functionApp \"{_projectFilePath}\"";
```
Comment (davidebbo) [inline Kudu.Core/Deployment/Generator/FunctionAppBuilder.cs]: So it's valid not to have a solution in the csproj case? I thought you said it wasn't.

File: Kudu.Core/Deployment/Generator/MicrosoftSiteBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+namespace Kudu.Core.Deployment.Generator
+{
+
+    public abstract class MicrosoftSiteBuilder : GeneratorSiteBuilder
+    {
+        private readonly string _projectFilePath;
+        private readonly string _solutionPath;
+        private readonly string _commandArgument;
+
+        protected string solutionPath
```
Comment (davidebbo) [inline Kudu.Core/Deployment/Generator/MicrosoftSiteBuilder.cs]: Casing should be SolutionPath. Also, use automatic prop and get rid of _solutionPath:

    protected string SolutionPath { get; private set; }

File: Kudu.Core/Deployment/Generator/MicrosoftSiteBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            get
+            {
+                return _solutionPath;
+            }
+        }
+
+        protected string commandArgument
```
Comment (davidebbo) [inline Kudu.Core/Deployment/Generator/MicrosoftSiteBuilder.cs]: Same as SolutionPath comment

File: Kudu.Core/Deployment/Generator/WebSiteBuilder.cs [hunk 5]
```diff
@@ -5,32 +5,31 @@
...
-            : base(environment, settings, propertyProvider, sourcePath)
+            : base(environment, settings, propertyProvider, sourcePath, null, solutionPath, "--aspWebSite")
         {
-            _projectPath = projectPath;
-            _solutionPath = solutionPath;
+            _projectPath = CleanPath(projectPath);
         }
 
+        // different from other MSSiteBuilders which does this "--aspWebSite {projectFilePath}",
+        // instead it uses "--aspWebSite --sitePath {projectPath}"
```
Comment (watashiSHUN) [inline Kudu.Core/Deployment/Generator/WebSiteBuilder.cs]: TODO separate it to 2 constructors, just like `functionAppBuilder` and `functionNodeBuilder` (when we use `--sitePath`, there's [no `msbuild` is involved](https://github.com/projectkudu/KuduScript/blob/master/lib/generator._js#L369-L373), only kudsync is called), which will also prevent me from using this ugly override method

File: Kudu.Core/Deployment/Generator/FunctionAppBuilder.cs [hunk 6]
```diff
@@ -2,18 +2,18 @@
...
+            : base(environment, settings, propertyProvider, sourcePath, projectFilePath, solutionPath, "--functionApp")
         {
         }
 
         public override string ProjectType
         {
             get
             {
-                return "functionApp";
+                return "functionApp Msbuild";
```
Comment (davidebbo) [inline Kudu.Core/Deployment/Generator/FunctionAppBuilder.cs]: Let's call this 'COMPILED FUNCTIONAPP'. (seems all the existing one are upper case)

File: Kudu.Core/Deployment/Generator/FunctionNodeBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
+﻿using Kudu.Contracts.Settings;
+
+namespace Kudu.Core.Deployment.Generator
+{
+    class FunctionNodeBuilder : BaseBasicBuilder
```
Comment (davidebbo) [inline Kudu.Core/Deployment/Generator/FunctionNodeBuilder.cs]: It's not really about Node, but about it not being compiled. e.g. if you use C# with .csx file, or PowerShell functions, you're still in that case

File: Kudu.Core/Deployment/Generator/FunctionNodeBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        public FunctionNodeBuilder(IEnvironment environment, IDeploymentSettingsManager settings, IBuildPropertyProvider propertyProvider, string repositoryPath, string projectPath)
+            : base(environment, settings, propertyProvider, repositoryPath, projectPath, "--functionApp")
+        {
+        }
+
+        public override string ProjectType
+        {
+            get
+            {
+                return "functionApp Node";
```
Comment (davidebbo) [inline Kudu.Core/Deployment/Generator/FunctionNodeBuilder.cs]: "NONCOMPILED FUNCTIONAPP"

File: Kudu.Core/Deployment/Generator/WebSiteBuilder.cs (hunk 5 above)
Comment (watashiSHUN) [inline Kudu.Core/Deployment/Generator/WebSiteBuilder.cs]: nvm...website is a special case, it doesn't take `projectFilePath` but it has `sitePath`

File: Kudu.Core/Deployment/Generator/CompileFunctionBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+        public CompileFunctionBuilder(IEnvironment environment, IDeploymentSettingsManager settings, IBuildPropertyProvider propertyProvider, string sourcePath, string projectFilePath, string solutionPath)
+            : base(environment, settings, propertyProvider, sourcePath, projectFilePath, solutionPath, "--functionApp")
+        {
+        }
+
+        public override string ProjectType
+        {
+            get
+            {
+                return "COMPILE FUNCTIONAPP";
```
Comment (davidebbo) [inline Kudu.Core/Deployment/Generator/CompileFunctionBuilder.cs]: 'COMPILE' -> 'COMPILED'

File: Kudu.Core/Deployment/Generator/NonCompileFunctionBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        public NonCompileFunctionBuilder(IEnvironment environment, IDeploymentSettingsManager settings, IBuildPropertyProvider propertyProvider, string repositoryPath, string projectPath)
+            : base(environment, settings, propertyProvider, repositoryPath, projectPath, "--functionApp")
+        {
+        }
+
+        public override string ProjectType
+        {
+            get
+            {
+                return "NONCOMPILE FUNCTIONAPP";
```
Comment (davidebbo) [inline Kudu.Core/Deployment/Generator/NonCompileFunctionBuilder.cs]: NONCOMPILED

File: Kudu.Core/Infrastructure/PathUtils/PathWindowsUtility.cs [hunk 11]
```diff
@@ -95,6 +95,12 @@ internal override string ResolveNpmJsPath()
             return Path.Combine(programFiles, "nodejs", npmCliPath);
         }
 
+        internal override string ResolveMSBuild15Dir()
+        {
+            string programFiles = SystemEnvironment.GetFolderPath(SystemEnvironment.SpecialFolder.ProgramFilesX86);
+            return Path.Combine(programFiles, "MSBuild-15.3-preview", "MSBuild", "15.0", "Bin");
```
Comment (davidebbo) [inline Kudu.Core/Infrastructure/PathUtils/PathWindowsUtility.cs]: We'll need to make sure to remember to change this once we get next preview.

File: Kudu.Core/Deployment/Generator/CompileFunctionBuilder.cs (hunk 9 above)
Comment (watashiSHUN) [inline Kudu.Core/Deployment/Generator/CompileFunctionBuilder.cs]: can I use the name in kuduscript: `MSBUILD FUNCTIONAPP`  and `BASIC FUNCTIONAPP`

File: Kudu.Core/Infrastructure/PathUtils/PathWindowsUtility.cs (hunk 11 above)
Comment (watashiSHUN) [inline Kudu.Core/Infrastructure/PathUtils/PathWindowsUtility.cs]: when I rebase, I can make it to be 2 commits, whereas one only has change related to **MSBuild15Dir**

**Pull request closed without merging**